### DISCUSSION TITLE
Does Lambda support CancellationToken?

### DISCUSSION THREAD

**Discussion by petro2050** [aws/aws-lambda-dotnet#1835] [Q&A] [answered]
I've noticed I implemented a lot of async code (services, repositories, etc.) with cancellation token support, but I'm actually not sure if Lambda supports this. Looking for an answer to help me decide if I should remove `CancellationToken` params from async methods.

Does `this.HttpContext.RequestAborted` work at controller level? For other functions, I've seen tricks like creating a token source from `lambdaContext.RemainingTime`.

**Comment by UnholyRanger** [comment, marked as answer]
If you need a cancellation token to pass around, you'll need to implement a CancellationSource in the handler method which will be based on the `LambdaContext.RemainingTime`. Once that time runs out, AWS will kill the lambda, so you need to subtract enough time to gracefully cancel. They have an example [here](https://docs.aws.amazon.com/lambda/latest/dg/csharp-context.html).
I've implemented an async job as well following the onion architecture (service-provider-repo) and the jobs could run <1 second or upwards of the entire time. I use a cancellation source that subtracts 2 minutes from the remaining time just to be safe and will adjust as needed.

**Comment by Dreamescaper** [comment]
> Does this.HttpContext.RequestAborted work at controller level?

No, AspNetCore lambda doesn't know whether request was aborted from client's side.
